src/WebJobs.Extensions.DurableTask/DurableOrchestrationContext.cs

```diff
@@ -397,9 +397,18 @@ internal void RaiseEvent(string name, string input)
                     Type tcsType = tcs.GetType();
                     Type genericTypeArgument = tcsType.GetGenericArguments()[0];
 
-                    object deserializedObject = MessagePayloadDataConverter.Default.Deserialize(input, genericTypeArgument);
-                    MethodInfo trySetResult = tcsType.GetMethod("TrySetResult");
-                    trySetResult.Invoke(tcs, new[] { deserializedObject });
+                    try
+                    {
+                        object deserializedObject = MessagePayloadDataConverter.Default.Deserialize(input, genericTypeArgument);
+                        MethodInfo trySetResult = tcsType.GetMethod("TrySetResult");
+                        trySetResult.Invoke(tcs, new[] { deserializedObject });
+                    }
+                    finally
+                    {
+                        // If we even attempt to raise an event, whether or not it failed, we should remove it from the pending collection
+                        // because otherwise WaitForExternal() will always find one with this key and run infinitely. Fixes #141
+                        this.pendingExternalEvents.Remove(name);

```

Comment:
`TaskCompletionSource.TrySetResult` will execute synchronously, so if the continuation wants to wait for another event with the same name, then you'll still get the infinite loop. I think you can fix this by removing it from the `pendingExternalEvents` list *first*, then calling `TrySetResult` on it.

Reply:
k, yeah i wondered about flipping the order as well. I'm in process of including a Unit Test for this alongside the checkin; will submit that change with it.

Thanks!